src/bin/options.ml

```diff
@@ -268,16 +269,183 @@ let report_style =
+}
+
+let parse_ext_opt s : (string * [ `Typing | `Model ] option) option =
+  match String.rindex s '.' with
+  | exception Not_found -> Some (s, None)
+  | pos ->
+    let extension_name = String.sub s 0 pos in
+    let plugin_kind = String.sub s (pos + 1) (String.length s - pos - 1) in
+    match plugin_kind with
+    | "typing" -> Some (extension_name, Some `Typing)
+    | "model" -> Some (extension_name, Some `Model)
+    | _ -> None
+
+let find_typing_plugin s name =
+  try
+    Builtin (
+      List.find (fun ext -> Dolmen_loop.Typer.Ext.name ext = name)
+        (Dolmen_loop.Typer.Ext.list ())
+    )
+  with Not_found -> Dune_plugin s
 
-let typing_ext =
-  Arg.enum typing_extension_list
+let find_model_plugin plugin name =
+  try
+    Builtin (
+      List.find (fun ext -> Dolmen_model.Ext.name ext = name)
+        (Dolmen_model.Ext.list ())
+    )
+  with Not_found -> Dune_plugin plugin
+
+let merge_extension e1 e2 =
+  assert (e1.extension_name = e2.extension_name);
+  let typing_plugin = merge_plugin e1.typing_plugin e2.typing_plugin
+  and model_plugin = merge_plugin e1.model_plugin e2.model_plugin in
+  { extension_name = e1.extension_name
+  ; typing_plugin
+  ; model_plugin }
+
+let merge_extensions extensions =
+  let tbl = Hashtbl.create 17 in
+  List.iter (fun ext ->
+    match Hashtbl.find tbl ext.extension_name with
+    | ext' ->
+      Hashtbl.replace tbl ext.extension_name (merge_extension ext ext')
+    | exception Not_found ->
+      Hashtbl.replace tbl ext.extension_name ext
+  ) extensions;
+  List.fast_sort (fun e1 e2 ->
+    String.compare e1.extension_name e2.extension_name
+  ) @@ Hashtbl.fold (fun _ e all -> e :: all) tbl []
+
+let builtin_typing_extensions =
+  List.map (fun ext ->
+    { extension_name = Dolmen_loop.Typer.Ext.name ext
+    ; typing_plugin = Builtin ext
+    ; model_plugin = Unavailable }
+  ) (Dolmen_loop.Typer.Ext.list ())
+
+let builtin_model_extensions =
+  List.map (fun ext ->
+    { extension_name = Dolmen_model.Ext.name ext
+    ; typing_plugin = Unavailable
+    ; model_plugin = Builtin ext }
+  ) (Dolmen_model.Ext.list ())
+
+let builtin_extensions =
+  merge_extensions
+    (List.rev_append builtin_typing_extensions builtin_model_extensions)
+
+let all_extensions, invalid_extensions =
+  let add_plugin ok err plugin = function
+    | None -> ok, plugin :: err
+    | Some (extension_name, k) ->
+      let typing_plugin =
+        match k with
+        | None | Some `Typing -> Dune_plugin plugin
+        | Some _ -> Unavailable
+      and model_plugin =
+        match k with
+        | None | Some `Model -> Dune_plugin plugin
+        | Some _ -> Unavailable
+      in
+      { extension_name ; typing_plugin ; model_plugin } :: ok, err
+  in
+  let ok, err =
+    List.fold_left (fun (ok, err) e ->
+      add_plugin ok err e @@ parse_ext_opt e
+    ) (builtin_extensions, []) (Dolmen.Sites.Plugins.Plugins.list ())
+  in
+  merge_extensions ok, List.fast_sort String.compare err
```

Comment:
Hm... I'm a bit worried by the cost of computing the list of all extensions systematically at startup (particularly since listing all dune plugins might involve a number of system calls).
Intuitively, we should only need that we want want to list all extensions, i.e. when a user asks for a list of extensions (or in error messages), but when loading an extension, we should be able to not need the lis tof all extensions.

Reply:
I think I did it this way because I initially wanted to preserve the behavior where the list of all available extensions is provided as an `enum` to Cmdliner. Since this is no longer the case (and there is `--list-extensions` instead), it makes sense to perform the listing lazily; will change.